File: documentapi/src/vespa/documentapi/messagebus/messages/batchdocumentupdatemessage.cpp
// Copyright 2016 Yahoo Inc. Licensed under the terms of the Apache 2.0 license. See LICENSE in the project root.

#include "batchdocumentupdatemessage.h"
#include "batchdocumentupdatereply.h"
#include <vespa/documentapi/messagebus/documentprotocol.h>
#include <vespa/document/bucket/bucketidfactory.h>
#include <vespa/vespalib/util/exceptions.h>

namespace documentapi {

BatchDocumentUpdateMessage::BatchDocumentUpdateMessage(uint64_t userId)
    : _userId(userId)
{
    setBucketId(document::UserDocIdString(vespalib::make_string("userdoc:foo:%lu:bar", _userId)));
}

BatchDocumentUpdateMessage::BatchDocumentUpdateMessage(const string& group)
    : _userId(0),
      _group(group)
{
    setBucketId(document::GroupDocIdString("groupdoc:foo:" + _group + ":bar"));
}

BatchDocumentUpdateMessage::~BatchDocumentUpdateMessage() {}

void
BatchDocumentUpdateMessage::setBucketId(const document::IdString& idString)
{
    document::BucketIdFactory factory;
    _bucketId = factory.getBucketId(document::DocumentId(idString));
}

void
BatchDocumentUpdateMessage::addUpdate(document::DocumentUpdate::SP update)
{
    verifyUpdate(*update);
    _updates.push_back(update);
}

void
BatchDocumentUpdateMessage::verifyUpdate(const document::DocumentUpdate& update) {
    const document::IdString& idString = update.getId().getScheme();

    if (_group.length()) {
        string group;

        if (idString.hasGroup()) {
            group = idString.getGroup();
        } else {
            throw vespalib::IllegalArgumentException("Batch update message can only contain groupdoc or orderdoc items");
        }

        if (group != _group) {
            throw vespalib::IllegalArgumentException(vespalib::make_string("Batch update message can not contain messages from group %s, only group %s", group.c_str(), _group.c_str()));
        }
    } else {
        uint64_t userId;

        if (idString.hasNumber()) {
            userId = idString.getNumber();
        } else {
            throw vespalib::IllegalArgumentException("Batch update message can only contain userdoc or orderdoc items");
        }

        if (userId != _userId) {
            throw vespalib::IllegalArgumentException(vespalib::make_string("Batch update message can not contain messages from user %llu, only user %llu", (long long unsigned)userId, (long long unsigned)_userId));
        }
    }
}

DocumentReply::UP
BatchDocumentUpdateMessage::doCreateReply() const
{
    return DocumentReply::UP(new BatchDocumentUpdateReply());
}

uint32_t
BatchDocumentUpdateMessage::getType() const {
    return DocumentProtocol::MESSAGE_BATCHDOCUMENTUPDATE;
}

}

File: documentapi/src/main/java/com/yahoo/documentapi/messagebus/protocol/BatchDocumentUpdateMessage.java
// Copyright 2017 Yahoo Holdings. Licensed under the terms of the Apache 2.0 license. See LICENSE in the project root.
package com.yahoo.documentapi.messagebus.protocol;

import com.yahoo.document.BucketId;
import com.yahoo.document.BucketIdFactory;
import com.yahoo.document.DocumentId;
import com.yahoo.document.DocumentUpdate;
import com.yahoo.document.idstring.GroupDocIdString;
import com.yahoo.document.idstring.IdString;
import com.yahoo.document.idstring.UserDocIdString;
import com.yahoo.document.serialization.DocumentDeserializer;

import java.nio.ByteBuffer;
import java.util.ArrayList;
import java.util.List;

/**
 * @author Simon Thoresen Hult
 */
public class BatchDocumentUpdateMessage extends DocumentMessage {

    private DocumentDeserializer buffer = null;
    private List<DocumentUpdate> updates = new ArrayList<DocumentUpdate>();
    private LazyDecoder decoder = null;
    private String group = null;
    private Long userId = null;
    private BucketId bucketId = null;

    public String getGroup() {
        return group;
    }

    public Long getUserId() {
        return userId;
    }

    /**
     * Constructs a new message for deserialization.
     */
    BatchDocumentUpdateMessage() {
        // empty
    }

    /**
     * Constructs a new message from a byte buffer.
     *  @param decoder The decoder to use for deserialization.
     * @param buffer  A byte buffer that contains a serialized message.
     */
    public BatchDocumentUpdateMessage(long userId, LazyDecoder decoder, DocumentDeserializer buffer) {
        this.userId = userId;
        this.decoder = decoder;
        this.buffer = buffer;
        setBucketId(new UserDocIdString("foo", userId, "bar"));
    }

    /**
     * Constructs a new message from a byte buffer.
     *  @param decoder The decoder to use for deserialization.
     * @param buffer  A byte buffer that contains a serialized message.
     */
    public BatchDocumentUpdateMessage(String group, LazyDecoder decoder, DocumentDeserializer buffer) {
        this.group = group;
        this.decoder = decoder;
        this.buffer = buffer;
        setBucketId(new GroupDocIdString("foo", group, "bar"));
    }

    /**
     * Constructs a batch document update message, to contain updates for documents for the given user.
     */
    public BatchDocumentUpdateMessage(long userId) {
        super();
        this.userId = userId;
        setBucketId(new UserDocIdString("foo", userId, "bar"));
    }

    /**
     * Constructs a batch document update message, to contain updates for documents for the given user.
     */
    public BatchDocumentUpdateMessage(String group) {
        super();
        this.group = group;
        setBucketId(new GroupDocIdString("foo", group, "bar"));
    }

    void setBucketId(IdString id) {
        BucketIdFactory factory = new BucketIdFactory();
        bucketId = factory.getBucketId(new DocumentId(id));
    }

    BucketId getBucketId() {
        return bucketId;
    }

    /**
     * This method will make sure that any serialized content is deserialized into proper message content on first
     * entry. Any subsequent entry into this function will do nothing.
     */
    private void deserialize() {
        if (decoder != null && buffer != null) {
            decoder.decode(this, buffer);
            decoder = null;
            buffer = null;
        }
    }

    /**
     * Returns the list of document updates to perform.
     *
     * @return The updates.
     */
    public List<DocumentUpdate> getUpdates() {
        deserialize();
        return updates;
    }

    /**
     * Adds a document update to perform.
     *
     * @param upd The document update to set.
     */
    public void addUpdate(DocumentUpdate upd) {
        buffer = null;
        decoder = null;

        verifyUpdate(upd);
        updates.add(upd);
    }

    /**
     * Returns the raw serialized buffer. This buffer is stored as the message is received from accross the network, and
     * deserialized from as soon as a member is requested. This method will return null if the buffer has been decoded.
     *
     * @return The buffer containing the serialized data for this message, or null.
     */
    ByteBuffer getSerializedBuffer() {
        return buffer != null ? buffer.getBuf().getByteBuffer() : null;
    }

    @Override
    public DocumentReply createReply() {
        return new BatchDocumentUpdateReply();
    }

    @Override
    public int getType() {
        return DocumentProtocol.MESSAGE_BATCHDOCUMENTUPDATE;
    }

    void verifyUpdate(DocumentUpdate update) {
        if (update == null) {
            throw new IllegalArgumentException("Document update can not be null.");
        }

        IdString idString = update.getId().getScheme();

        if (group != null) {
            String idGroup;

            if (idString.hasGroup()) {
                idGroup = idString.getGroup();
            } else {
                throw new IllegalArgumentException("Batch update message can only contain groupdoc or orderdoc items");
            }

            if (!group.equals(idGroup)) {
                throw new IllegalArgumentException("Batch update message can not contain messages from group " + idGroup + " only group " + group);
            }
        } else {
            long foundUserId = 0;

            if (idString.hasNumber()) {
                foundUserId = idString.getNumber();
            } else {
                throw new IllegalArgumentException("Batch update message can only contain userdoc or orderdoc items");
            }

            if (userId != foundUserId) {
                throw new IllegalArgumentException("Batch update message can not contain messages from user " + foundUserId + " only user " + userId);
            }
        }
    }

}
